byoc/byoc.go

```diff
@@ -46,38 +37,141 @@ func NewBYOCGatewayServer(node *core.LivepeerNode, statusStore StatusStore, whip
 		sharedBalMtx: &sync.Mutex{},
 	}
 
+	bsg.StreamPipelines = make(map[string]*BYOCStreamPipeline)
+
 	bsg.registerRoutes()
 	return bsg
 }
 
-func NewBYOCOrchestratorServer(node *core.LivepeerNode, orch Orchestrator, trickleSrv *trickle.Server, trickleBasePath string, mux *http.ServeMux) *BYOCOrchestratorServer {
-	bso := &BYOCOrchestratorServer{
-		node:            node,
-		orch:            orch,
-		trickleSrv:      trickleSrv,
-		trickleBasePath: trickleBasePath,
-		httpMux:         mux,
-		sharedBalMtx:    &sync.Mutex{},
-	}
-
-	bso.registerRoutes()
-	return bso
-}
-
 func (bsg *BYOCGatewayServer) Node() *core.LivepeerNode {
 	return bsg.node
 }
 func (bsg *BYOCGatewayServer) SharedBalanceLock() *sync.Mutex {
 	return bsg.sharedBalMtx
 }
 
+func (bsg *BYOCGatewayServer) newStreamPipeline(requestID, streamID, pipeline string, streamParams byocAIRequestParams, streamRequest []byte) *BYOCStreamPipeline {
+	streamCtx, streamCancel := context.WithCancelCause(context.Background())
+	bsg.mu.Lock()
+	defer bsg.mu.Unlock()
+
+	//ensure streamRequest is not nil or empty to avoid json unmarshal issues on Orchestrator failover
+	//sends the request bytes to next Orchestrator
+	if len(streamRequest) == 0 {
+		streamRequest = []byte("{}")
+	}
+
+	bsg.StreamPipelines[streamID] = &BYOCStreamPipeline{
+		RequestID:     requestID,
+		StreamID:      streamID,
+		Pipeline:      pipeline,
+		streamCtx:     streamCtx,
+		streamParams:  streamParams,
+		streamCancel:  streamCancel,
+		streamRequest: streamRequest,
+		OutCond:       sync.NewCond(bsg.mu),
+	}
+	return bsg.StreamPipelines[streamID]
+}
+
+func (bsg *BYOCGatewayServer) streamPipeline(streamId string) (*BYOCStreamPipeline, error) {
+	bsg.mu.Lock()
+	defer bsg.mu.Unlock()
+	p, exists := bsg.StreamPipelines[streamId]
+	if !exists {
+		return nil, fmt.Errorf("BYOC Stream pipeline %s not found", streamId)
+	}
+	return p, nil
+}
+
+func (bsg *BYOCGatewayServer) streamPipelineExists(streamId string) bool {
+	bsg.mu.Lock()
+	defer bsg.mu.Unlock()
+	_, exists := bsg.StreamPipelines[streamId]
+	return exists
+}
+
+func (bsg *BYOCGatewayServer) stopStreamPipeline(streamId string, err error) {
+	p, err := bsg.streamPipeline(streamId)
+	if err == nil {
+		p.OutCond.Broadcast()
```

Comment:
It looks like we broadcast and flip pipeline state without holding the cond’s mutex, so waiters could miss the signal or see inconsistent state during shutdown. Could we wrap the Closed flag and OutCond.Broadcast() under the cond lock, then release the lock before doing the slower close calls?
```suggestion
func (bsg *BYOCGatewayServer) stopStreamPipeline(streamId string, err error) {
	stream, serr := bsg.streamPipeline(streamId)
	if serr != nil {
		return
	}

	stream.OutCond.L.Lock()
	stream.Closed = true
	stream.OutCond.Broadcast()
	ctrlPub := stream.ControlPub
	stopCtrl := stream.StopControl
	stream.OutCond.L.Unlock()

	if ctrlPub != nil {
		if cerr := ctrlPub.Close(); cerr != nil {
			glog.Errorf("Error closing trickle publisher: %v", cerr)
		}
	}
	if stopCtrl != nil {
		stopCtrl()
	}
	stream.streamCancel(err)
}
```